PR: https://github.com/odoo/odoo/pull/

From: 98f974cae305f868de97682db8b36e5c570d9e70
From: Simon Lejeune
Date: 2019-08-07 14:11:21

Structural Changes: 4
Total Changes: 40

[REF] stock_landed_costs: default journal id

Set a default journal on the company, use it on the landed costs form.
I was not able to directly set the value on the filed with a post init
hook because the stock journal is generated a while after the landed
costs installation in the l10n_generic_coa module.

task-59011

================================= pseudo patch: =================================

--- a/addons/stock_landed_costs/models/__init__.py
+++ b/addons/stock_landed_costs/models/__init__.py
@@ -2,6 +2,8 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from . import product
+from . import res_company
+from . import res_config_settings
 from . import stock_landed_cost
 from . import stock_valuation_layer
 

--- a/None
+++ b/addons/stock_landed_costs/models/res_company.py
@@ -0,0 +1,11 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = 'res.company'
+
+    lc_journal_id = fields.Many2one('account.journal')
+

--- a/None
+++ b/addons/stock_landed_costs/models/res_config_settings.py
@@ -0,0 +1,11 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class ResConfigSettings(models.TransientModel):
+    _inherit = 'res.config.settings'
+
+    lc_journal_id = fields.Many2one('account.journal', string='Default Journal', related='company_id.lc_journal_id', readonly=False)
+

--- a/addons/stock_landed_costs/models/stock_landed_cost.py
+++ b/addons/stock_landed_costs/models/stock_landed_cost.py
@@ -14,6 +14,20 @@ class LandedCost(models.Model):
     _description = 'Stock Landed Cost'
     _inherit = ['mail.thread', 'mail.activity.mixin']
 
+    def _default_account_journal_id(self):
+        """Take the journal configured in the company, else fallback on the stock journal."""
+        lc_journal = self.env['account.journal']
+        if self.env.company.lc_journal_id:
+            lc_journal = self.env.company.lc_journal_id
+        else:
+            ir_property = self.env['ir.property'].search([
+                ('name', '=', 'property_stock_journal'),
+                ('company_id', '=', self.env.company.id)
+            ], limit=1)
+            if ir_property:
+                lc_journal = ir_property.get_by_record()
+        return lc_journal
+
     name = fields.Char(
         'Name', default=lambda self: _('New'),
         copy=False, readonly=True, tracking=True)
@@ -44,7 +58,7 @@ class LandedCost(models.Model):
         copy=False, readonly=True)
     account_journal_id = fields.Many2one(
         'account.journal', 'Account Journal',
-        required=True, states={'done': [('readonly', True)]})
+        required=True, states={'done': [('readonly', True)]}, default=lambda self: self._default_account_journal_id())
     company_id = fields.Many2one('res.company', string="Company",
         related='account_journal_id.company_id', readonly=False)
     stock_valuation_layer_ids = fields.One2many('stock.valuation.layer', 'stock_landed_cost_id')
